### ISSUE TITLE
Document/implement go routine & channel guarantees

### ISSUE THREAD

**Issue by FlorianUekermann** [gopherjs/gopherjs#720]
I am using the following endless loop in a go routine:
```
for {		
	select {
	case a = channelA:
		doSomething(a)
	case b = channelB:
		doSomething(b)
	case channelC <- 1:
	}
	doSomethingD()
}
```
I suspect that sending values to channelA or channelB is guaranteed to be non-blocking, as long as all the doSomethingX functions contain no blocking code. By non-blocking I mean that channel sends from event listeners are guaranteed to to not panic.

Is my assumption correct? I couldn't find any documentation. If it isn't correct, I think there would be value in implementing such a guarantee.

**Comment by theclapp** [comment]
I assume you mean this?
```go
// ...
case a := <- channelA:
// etc ...
```

I, for one, do not understand your question.  If doSomethingD() is running, then sends on channelA and channelB will block, unless those channels are buffered.  Perhaps I don't understand your clarification about not panicking.

**Comment by FlorianUekermann** [comment]
I'm using blocking in a very gopherjs specific sense here. Maybe there is a better word.
In gopherjs channel sends/receives will panic in if the code is called from js (most commonly event listener callbacks), unless a go routine is already waiting to receive/send on the same channel.

Consider this piece of code:
```
js.Global.Get("myButton").Call("addEventListener", "click", func() {
	channelA <- 1
})
```
If you click on the corresponding button and there is no go routine ready to receive on channelA, you will get this error and a stack trace (not sure if it is technically a panic):
```
Uncaught Error: runtime error: cannot block in JavaScript callback, fix by wrapping code in goroutine
```
However, there is a running go routine consisting of the for loop above, it may already have advanced to the select statement, which would prevent the error. But I don't know if this is actually guaranteed.

I other words, I am asking if there is a guarantee that said go routine will advance to the select before any other code is executed, both after starting the go routine and every send/receive inside the select.

**Comment by theclapp** [comment]
> I[n] other words, I am asking if there is a guarantee that said go routine will advance to the select 

I don't think there is.  Consider if `doSomethingD()` was just `time.Sleep(10*time.Second)`.  I think you could expect JS event handlers to run during that time, but the `select` would not be running.

The usual approach is to put the send to the channel in a new goroutine, as recommended in the error message:

```go
js.Global.Get("myButton").Call("addEventListener", "click", func() {
        go func() {
                channelA <- 1
        }()
})
```

or use a buffered channel.

Ah, but you said "as long as all the doSomethingX functions contain no blocking code", and `time.Sleep()` blocks.

In that case, well, I don't know for sure, but I don't think that's guaranteed, and if it is, it's probably an accidental implementation detail, not a formal commitment from the maintainers.

Any particular reason you don't want to put the channel send in a goroutine?

**Comment by shurcooL** [comment]
> But given the single threaded execution of javascript and the implementation details of gopherjs, it may actually be guaranteed that any non-blocking doSomethingX() and all other code outside the select finish executing before other code is run.

I would describe it as an implementation detail, not a hard guarantee. It may work now, it may or may not work later. GopherJS doesn't try to make additional guarantees about this beyond the Go language specification. You need to decide if that works for your use case.

Using a buffered channel (as long as buffered channel doesn't get full) or goroutine, as @theclapp suggested, would give you a guarantee the code won't block.

Since you're talking about this in the context of UI event processing patterns, I wanted to share a neat pattern I discovered recently. In my situation, I have a button that has effect only some of the time. Other time it has no effect. I wrote:

```Go
channelA = make(chan int)
js.Global.Get("myButton").Call("addEventListener", "click", func() {
	select {
	case channelA <- 1:
	default:
	}
})
```

By wrapping the channel send in a select with a default case, I have a guarantee that code is not blocking.

https://github.com/shurcooL/tictactoe/commit/2233059014b0c53f1f7a1bf2b09019f5ecf5961c is where I used this pattern, if you want see it in context.

**Comment by FlorianUekermann** [comment]
Now that I've had this in the back of my head for a day, I think I have found a less convoluted way of thinking about this. But let me answer your replies first.

Reply to @theclapp:
> ```
> js.Global.Get("myButton").Call("addEventListener", "click", func() {
>        go func() {
>                channelA <- 1
>        }()
> })
> ```

That is what I usually go for if ordering doesn't matter. However, ordering does often matter. There are ways to deal with this by queuing up actions and moving the ordering guarantee into the go routine. But it is by no means trivial, nor elegant and idiomatic.
Also note that for events that bubble and trigger multiple event listeners the ordering problem is far from a theoretical issue.

Reply to @shurcooL:
> ```
> js.Global.Get("myButton").Call("addEventListener", "click", func() {
>	select {
>	case buttonClick <- struct{}{}:
>	default:
>	}
> })
> ```

That is what I do on most clicks. But for other events, neither this pattern nor the one above (if ordering matters) are sufficient. Take change events as an example, you can't necessarily afford to skip a change event. (And I guess we can all agree that it would be great to make this pattern unnecessary.)
> GopherJS doesn't try to make additional guarantees about this beyond the Go language specification.

This doesn't go beyond the go spec. On the contrary, I'm trying to narrow down the scope of the necessary Go spec is violations JS forces us to live with.

--------

With the replies out of the way let me try to formulate a more straight forward and more general guarantee. To make this easier let me define a term:
**transient deadlock** - A situation where no code is ready to be run, because the runtime is waiting for the completion of an operation that is asynchronous in JS. Compared to an permanent deadlock this includes any situation where all go routines are temporarily stuck on channel sends/receives that can't proceed, time.Sleep(), waiting for sync.Mutex.Lock(), network IO etc.

If the runtime encounters a permanent or transient deadlock in Go code called from JS (like JS callbacks) you will get `runtime error: cannot block in JavaScript callback` or similar. This is not only the current behavior, but also seems unavoidable.

But it would be great if GopherJS would guarantee that these errors only occur if there is a permanent or transient deadlock. I think this could be accomplished in practice by running any go routine that can proceed before crashing with those `runtime error: cannot block` errors.

As a result all code that avoids calling time.Sleep(), http.Get(), http.Post() (I'm sure I forgot a bunch) would be guaranteed to work just fine, even if it used sync.Mutex.Lock and channels inside JS callbacks.

From what I think I know about the current implementation this may actually be the case already, or at least be relatively straightforward. Also, I'm sure I missed some details, but I hope I made the overall idea sufficiently clear.

**Comment by FlorianUekermann** [comment]
I have spent some time reading compiler/prelude/goroutines.go and am more confident now that the approach outlined in my original question is safe. I also ran some tests:
```
func main() {
	js.Global.Get("document").Call("addEventListener", "DOMContentLoaded", func() {
		var c = make(chan int)
		go func() {
			for i := 0; true; i++ {
				println("receiving", i)
				var i = <-c
				println("received", i)
			}
		}()
		for i := 0; i < 2; i++ {
			println("sending", i)
			c <- i
			println("sent", i)
		}
	})
}
```

Output:
```
receiving 0
sending 0
received 0
receiving 1
sent 0
sending 1
received 1
receiving 2
sent 1
```

Conclusion:
(By "synchronized" operations I mean stuff that may suspend go routine execution.)
Go routines are executed immediately on dispatch and will proceed to the first statement that requires suspension of execution. So the first send/receive to/from any dispatched go routine is safe from a JS callback, if there are no other  receive is the first "synchronized" operation.
Furthermore, the go routine will not only receive/send the value as soon as the callback requires it, but also resume execution and continues to the next "synchronized" operation, before the callback proceeds.
Tl;dr: If you have an go routine draining/feeding values from/to your callbacks, you don't need to worry about using channels in them, as long as you don't do stuff like sleep or IO in between.

If I can find the time I still want to verify or implement the "don't crash unless there is no other choice"
 behavior I described in the previous comment "transient deadlock. It would be nice to tell users: "if it can proceed, it will", because that is much easier to reason about.

**Comment by FlorianUekermann** [comment]
One last update, before I would like to wait for @neelance take on this, since it doesn't seem like anyone else touched that logic.

The scheduling code is in compiler/prelude/goroutines.go. Making sure that we are in a deadlock scenario and if not scheduling the execution of some go routine would start around line 211 (for sending) and 240 (for receiving). All other operations that suspend execution I could find (I looked at the obvious candidates from sync, net/http, time) are built on regular channels.

However, before writing that code, I would like to find a test case, that actually throws `Uncaught Error: runtime error: cannot block in JavaScript callback...` in a case without a (transient) deadlock.
If it turns out that such errors can't happen without deadlock, I propose changing the error message and documentation on this topic accordingly, so people can start using channels in callbacks more liberally.

**Comment by theclapp** [comment]
> However, ordering does often matter.

Understood.

> There are ways to deal with this by maintaining an action queue and moving the ordering guarantee into the go routine. But the general case is by no means trivial, nor elegant and idiomatic.
> Also note that for events that bubble and trigger multiple event listeners the ordering problem is far from a theoretical issue.

I have a question about this.  (Mind you, I'm _not_ saying it's wrong!)  I don't have tons of front-end experience.  Can you elaborate on what's wrong with using a synchronous send (_not_ in its own goroutine) to a channel with a sufficiently large buffer?  Not having had to deal with this challenge myself, it seems like this would address the ordering concern in a trivial, elegant, idiomatic fashion.  It seems like anything else you do on the receiver would boil down to "read the event from the channel and put it on an internal queue" ... so why not just let the channel be the queue?

If this is a big question to answer, I'd be happy to create a different issue so as not to derail the primary discussion here.  Thank you for your help.

**Comment by FlorianUekermann** [comment]
Sorry for the delay, I was travelling. I'll reply in thread and with a little more text than necessary, because I think the answer underlines the practical importance of documenting this kind of guarantee (I'm pretty sure by now that the current scheduler is already as spec compliant as possible, until JS implements "sleep").

> Can you elaborate on what's wrong with using a synchronous send (not in its own goroutine) to a channel with a sufficiently large buffer?

Ordering. Consider a slightly more general scenario than the one you describe (like the one in the OP). Once channels are buffered the language doesn't guarantee that a select statement receives/sends in order.

And even if you ignore ordering. Think about channels or locks in a websocket callback:
1. The user can only expect what the go spec guarantees or is documented. Therefore the average user doesn't know that *infinite* channel capacities (e.g. 2<<31) are often ok in gopherjs.
2. Without any guarantees beyond the spec, a buffered send may even be a bug, because it may lead to DOS situations with very frequent callers (no time to call the processing routine) or even out of memory errors. This may not apply to UI scenarios, because users may not click that often, but the other end of a websocket could be feeding you many GBs of data at high speed.

In general, there are very good reasons why go has unbuffered channels and those reasons are as valid in gopherjs as they are everywhere else. It's usually boils down to ordering or resource exhaustion.

There is also the more practical issue, that you may want to use libraries in your callback without forking them every time they use a lock or channel or wrapping their uses in complex workarounds.

**Comment by theclapp** [comment]
Thank you for your comprehensive response.  So, thinking out loud:

> Ordering. Consider a slightly more general scenario than the one you describe (like the one in the OP). Once channels are buffered the language doesn't guarantee that a select statement receives/sends in order.

So your point is that in the case of your original code

```go
for {		
	select {
	case a := <-channelA:
		doSomething(a)
	case b := <-channelB:
		doSomething(b)
	case channelC <- 1:
	}
	doSomethingD()
}
```

using buffered channels, there's no guarantee that channelA and channelB will be read from in the same order they were sent to.  E.g. if you send on A then B then A then B, you might well receive A then A then B then B.

And sure, you could make A and B unbuffered and have a goroutine that reads from them both and writes to a single buffered channel to maintain ordering, but a) that is probably still insufficient in many cases, and b) it certainly starts to get non-trivial and inelegant, which was your point.  (Or you could combine A and B into a single (buffered) channel, but then you probably need typeswitches on the receive side to figure out what you got, and you're back to non-trivial and inelegant and probably still insufficient to boot.)

Thanks again for your response and for bearing with me while I work my way through understanding it.

(Also, ICYDK, if you start your code blocks with ` ```go` then you get Go syntax highlighting.)
